### PR TITLE
feat: Add a new dataflow template to verify integrity of data imported from HBase.

### PR DESCRIPTION
**Pull request by vermas2012**
[DRAFT] A new beam pipeline to validate the integrity of data imported from HBase.

### CODE REVIEW

File: bigtable-dataflow-parent/bigtable-beam-import/src/main/resources/log4j.properties [hunk 1]
```diff
@@ -22,3 +22,5 @@ log4j.appender.stdout.layout=org.apache.log4j.PatternLayout
 log4j.appender.stdout.layout.ConversionPattern=%d{yyyy-MM-dd HH:mm:ss} %-5p %c{1}:%L - %m%n
 
 log4j.category.org.apache.hadoop.io.compress.CodecPool=WARN
+log4j.logger.org.apache.hadoop.io.compress.Compression=TRACE
+log4j.logger.org.apache.hadoop.util.NativeCodeLoader=DEBUG
```
Comment (igorbernstein2) [inline bigtable-dataflow-parent/bigtable-beam-import/src/main/resources/log4j.properties:26]: are these levels intentional?

File: bigtable-dataflow-parent/bigtable-beam-import/src/test/java/com/google/cloud/bigtable/beam/validation/HadoopHashBasedReaderTest.java [hunk 2]
```diff
@@ -0,0 +1,280 @@
...
+import org.apache.hadoop.conf.Configuration;
+import org.apache.hadoop.hbase.HConstants;
+import org.apache.hadoop.hbase.io.ImmutableBytesWritable;
+import org.apache.hadoop.hbase.mapreduce.HashTable.TableHash;
+import org.junit.After;
+import org.junit.Before;
+import org.junit.Test;
+import org.mockito.Mockito;
+
+public class HadoopHashBasedReaderTest extends TestCase {
```
Comment (igorbernstein2) [inline bigtable-dataflow-parent/bigtable-beam-import/src/test/java/com/google/cloud/bigtable/beam/validation/HadoopHashBasedReaderTest.java:41]: Can you take a look at the tooling that beam provides for testing custom sources and see if you can utilize them here? (same for the other test)

Also `RunWith(Junit4.class)`

File: pom.xml [hunk 3]
```diff
@@ -76,10 +76,10 @@ limitations under the License.
     <hamcrest.version>1.3</hamcrest.version>
     <mockito.version>3.6.28</mockito.version>
     <!-- TODO: check if commons-codec was transitively updated to 1.13 and okhttp was updated to 2.7.5 when upgrading-->
-    <beam.version>2.19.0</beam.version>
+    <beam.version>2.24.0</beam.version>
     <!-- referred from bigtable-beam-import and bigtable-emulator -->
     <guava.version>30.0-android</guava.version>
-    <beam-guava.version>20.0</beam-guava.version>
+    <beam-guava.version>29.0-jre</beam-guava.version>
```
Comment (igorbernstein2) [inline pom.xml:82]: please extract version bumps to separate pr

**Review by igorbernstein2 — CHANGES_REQUESTED** [review]
I took a rough pass reviewing. I'll probably have more comments after the holidays, but I wanted to jot some initial thoughts

File: bigtable-dataflow-parent/bigtable-beam-import/pom.xml [hunk 4]
```diff
@@ -78,6 +76,21 @@ limitations under the License.
       <version>${hbase.version}</version>
     </dependency>
 
+    <!-- For HBase 2.x, this should be hbase-mapreduce
+    https://hbase.apache.org/2.1/book.html#export
+    -->
+    <dependency>
```
[the remark this answers is not in the document]
Comment (vermas2012) [reply bigtable-dataflow-parent/bigtable-beam-import/pom.xml:82]: This PR has all the changes from the import snapshot PR but is comparing against the master. The fix will be made to import snapshot and I will merge it before pushing.

File: bigtable-dataflow-parent/bigtable-beam-import/src/main/java/com/google/cloud/bigtable/beam/validation/ComputeAndValidaeHashFromBigtableDoFn.java [hunk 5]
```diff
@@ -0,0 +1,209 @@
...
+import org.apache.hadoop.hbase.client.Table;
+import org.apache.hadoop.hbase.io.ImmutableBytesWritable;
+import org.apache.hadoop.hbase.mapreduce.BigtableTableHashAccessor.BigtableResultHasher;
+import org.apache.hadoop.hbase.util.Bytes;
+
+/**
+ * A {@link DoFn} that takes a row range and hash from HBase and validates the hash from rows read
+ * from Cloud Bigtable.
+ */
+public class ComputeAndValidaeHashFromBigtableDoFn
```
[the remark this answers is not in the document]
Comment (vermas2012) [reply bigtable-dataflow-parent/bigtable-beam-import/src/main/java/com/google/cloud/bigtable/beam/validation/ComputeAndValidaeHashFromBigtableDoFn.java:43]: Done.

File: bigtable-dataflow-parent/bigtable-beam-import/src/main/java/com/google/cloud/bigtable/beam/validation/ComputeAndValidaeHashFromBigtableDoFn.java [hunk 6]
```diff
@@ -0,0 +1,209 @@
...
+
+  private transient BigtableSession session;
+
+  public ComputeAndValidaeHashFromBigtableDoFn(SyncTableOptions options) {
+    // TODO Remove the dependency on import options after ImportFromSnapshot is committed.
+    super(TemplateUtils.BuildImportConfig(options));
+    this.tableName = options.getBigtableTableId().get();
+  }
+
+  private String immutableBytesToString(ImmutableBytesWritable bytes) {
```
[the remark this answers is not in the document]
Comment (vermas2012) [reply bigtable-dataflow-parent/bigtable-beam-import/src/main/java/com/google/cloud/bigtable/beam/validation/ComputeAndValidaeHashFromBigtableDoFn.java:57]: done

File: bigtable-dataflow-parent/bigtable-beam-import/src/main/java/com/google/cloud/bigtable/beam/validation/ComputeAndValidaeHashFromBigtableDoFn.java [hunk 7]
```diff
@@ -0,0 +1,209 @@
...
+
+/**
+ * A {@link DoFn} that takes a row range and hash from HBase and validates the hash from rows read
+ * from Cloud Bigtable.
+ */
+public class ComputeAndValidaeHashFromBigtableDoFn
+    extends AbstractCloudBigtableTableDoFn<KV<String, Iterable<List<RangeHash>>>, RangeHash> {
+
+  private static final long serialVersionUID = 1L;
+  private String tableName;
```
[the remark this answers is not in the document]
Comment (vermas2012) [reply bigtable-dataflow-parent/bigtable-beam-import/src/main/java/com/google/cloud/bigtable/beam/validation/ComputeAndValidaeHashFromBigtableDoFn.java:47]: done

File: bigtable-dataflow-parent/bigtable-beam-import/src/main/java/com/google/cloud/bigtable/beam/validation/SyncTableJob.java [hunk 8]
```diff
@@ -0,0 +1,194 @@
+/*
+ * Copyright 2017 Google Inc. All Rights Reserved.
```
[the remark this answers is not in the document]
Comment (vermas2012) [reply bigtable-dataflow-parent/bigtable-beam-import/src/main/java/com/google/cloud/bigtable/beam/validation/SyncTableJob.java:2]: done

File: bigtable-dataflow-parent/bigtable-beam-import/src/main/java/com/google/cloud/bigtable/beam/validation/ComputeAndValidaeHashFromBigtableDoFn.java [hunk 9]
```diff
@@ -0,0 +1,209 @@
...
+
+      // Since all the row-ranges are sorted in HashTable's data files, 1 big scan can be used
+      // to read all the row ranges. Parallelism is achieved by splitting the HashTable's data
+      // files into smaller bundle of row-ranges by GroupBy.
+      ResultScanner scanner = createBigtableScan(rangeStartInclusive, rangeEndExclusive);
+
+      int bufferIndex = 0;
+      // TODO remove bufferIndex and use the iterator
+      Iterator<RangeHash> bufferIter = buffer.iterator();
+      int numRows = 0;
```
[the remark this answers is not in the document]
Comment (vermas2012) [reply bigtable-dataflow-parent/bigtable-beam-import/src/main/java/com/google/cloud/bigtable/beam/validation/ComputeAndValidaeHashFromBigtableDoFn.java:99]: done.

File: bigtable-dataflow-parent/bigtable-beam-import/src/main/java/com/google/cloud/bigtable/beam/validation/ComputeAndValidaeHashFromBigtableDoFn.java [hunk 10]
```diff
@@ -0,0 +1,209 @@
+/*
+ * Copyright 2017 Google Inc. All Rights Reserved.
```
[the remark this answers is not in the document]
Comment (vermas2012) [reply bigtable-dataflow-parent/bigtable-beam-import/src/main/java/com/google/cloud/bigtable/beam/validation/ComputeAndValidaeHashFromBigtableDoFn.java:2]: done

File: bigtable-dataflow-parent/bigtable-beam-import/src/main/java/com/google/cloud/bigtable/beam/validation/GenerateGroupByKeyDoFn.java [hunk 11]
```diff
@@ -0,0 +1,79 @@
...
+    buffer.add(context.element());
+    // Used to distribute work amongst workers. We want to split a file but want to retain
+    // the contiguous nature of scans. So that for each bundle we can do a single bigtable scan.
+
+    if (buffer.size() == batchSize) {
+      // This batch is complete, next item will start a new batch.
+      context.output(KV.of(Bytes.toStringBinary(buffer.get(0).startInclusive), buffer));
+      buffer = new ArrayList<>(batchSize);
+    }
+  }
```
[the remark this answers is not in the document]
Comment (vermas2012) [reply bigtable-dataflow-parent/bigtable-beam-import/src/main/java/com/google/cloud/bigtable/beam/validation/GenerateGroupByKeyDoFn.java:78]: Good catch ! done.

File: bigtable-dataflow-parent/bigtable-beam-import/src/main/java/com/google/cloud/bigtable/beam/validation/GenerateGroupByKeyDoFn.java [hunk 12]
```diff
@@ -0,0 +1,79 @@
...
+ */
+public class GenerateGroupByKeyDoFn extends DoFn<RangeHash, KV<String, List<RangeHash>>> {
+
+  // Batch 50 row ranges in a batch. We may change it in future to be based on the size of a batch.
+  // Having smaller batches also helps in validateHash stage as it doesn't emit an output in happy
+  // case. Finishing smaller bundles will tell the dataflow service that worker is alive.
+  private static final int DEFAULT_BATCH_SIZE = 50;
+  private static final long serialVersionUID = 1L;
+  private final int batchSize;
+  private List<RangeHash> buffer;
```
[the remark this answers is not in the document]
Comment (vermas2012) [reply bigtable-dataflow-parent/bigtable-beam-import/src/main/java/com/google/cloud/bigtable/beam/validation/GenerateGroupByKeyDoFn.java:45]: yes.

File: bigtable-dataflow-parent/bigtable-beam-import/src/main/java/com/google/cloud/bigtable/beam/validation/GenerateGroupByKeyDoFn.java [hunk 13]
```diff
@@ -0,0 +1,79 @@
...
+/**
+ * A {@link DoFn} function that takes a range and hash from HBase and generates a batch or work. The
+ * key is the start key of the batch and value is the sorted range of RangeHashes. This function
+ * generates a KV<String, List<RangeHash>> to preserve the sorted nature of RangeHashes. There is no
+ * guarantee of preserving the order of RangeHash if this function returns KV<String,RangeHash>.
+ *
+ * <p>Sorted list of RangeHashes allows the subsequent functions to issue a single scan to Bigtable
+ * covering all the ranges in the list of RangeHash.
+ */
+public class GenerateGroupByKeyDoFn extends DoFn<RangeHash, KV<String, List<RangeHash>>> {
```
[the remark this answers is not in the document]
Comment (vermas2012) [reply bigtable-dataflow-parent/bigtable-beam-import/src/main/java/com/google/cloud/bigtable/beam/validation/GenerateGroupByKeyDoFn.java:37]: > have your source emit KV<partition_start_key, RangeHash>
We can't do that since we want to maintain the order of RangeHash. Beam does not guarantee the relative ordering of elements after groupby. I have tested this. 

> use GroupIntoBatches to create smaller bundles
GroupInToBatches requires a batch key to start with. We don't have that, so even with GroupInToBatches we need to iterate over the bundle and generate the keys. I prefer to keep it simple and group linearly (we will never have multiple keys to bundle).

[the remark this answers is not in the document]
Comment (vermas2012) [reply bigtable-dataflow-parent/bigtable-beam-import/src/main/java/com/google/cloud/bigtable/beam/validation/GenerateGroupByKeyDoFn.java:37]: done.

File: bigtable-dataflow-parent/bigtable-beam-import/src/main/java/com/google/cloud/bigtable/beam/validation/HadoopHashTableSource.java [hunk 14]
```diff
@@ -0,0 +1,411 @@
...
+  @Override
+  public BoundedReader<RangeHash> createReader(PipelineOptions options) {
+    // Reader should always come from KeyBasedSource.
+    throw new IllegalArgumentException("Reader can't be created from HadoopHashTableSource.");
+  }
+
+  // A Beam source to read from a single workitem.
+  protected static class KeyBasedHashTableSource extends HadoopHashTableSource {
+
+    private final String startRow;
```
[the remark this answers is not in the document]
Comment (vermas2012) [reply bigtable-dataflow-parent/bigtable-beam-import/src/main/java/com/google/cloud/bigtable/beam/validation/HadoopHashTableSource.java:213]: Source must be serializable and having immutable bytes writable makes this source not serializable.

File: bigtable-dataflow-parent/bigtable-beam-import/src/main/java/com/google/cloud/bigtable/beam/validation/HadoopHashTableSource.java [hunk 15]
```diff
@@ -0,0 +1,411 @@
...
+
+  @Override
+  public Coder<RangeHash> getOutputCoder() {
+    return coder;
+  }
+
+  @Override
+  public long getEstimatedSizeBytes(PipelineOptions options) throws Exception {
+    // HashTable data files don't expose a method to estimate size or lineCount.
+    return 0;
```
[the remark this answers is not in the document]
Comment (vermas2012) [reply bigtable-dataflow-parent/bigtable-beam-import/src/main/java/com/google/cloud/bigtable/beam/validation/HadoopHashTableSource.java:201]: The size of files doesn't really give us any information. This source only generates the hashes, the real data is coming from CBT.

File: bigtable-dataflow-parent/bigtable-beam-import/src/main/java/com/google/cloud/bigtable/beam/validation/HadoopHashTableSource.java [hunk 16]
```diff
@@ -0,0 +1,411 @@
...
+          "Starting a new reader at key range ["
+              + immutableBytesToString(startRow)
+              + " ,"
+              + immutableBytesToString(stopRow)
+              + ").");
+      numKeys = 0;
+
+      if (reader.next()) {
+        cachedBatchStartKey = reader.getCurrentKey();
+        if (cachedBatchStartKey.compareTo(stopRow) >= 0) {
```
[the remark this answers is not in the document]
Comment (vermas2012) [reply bigtable-dataflow-parent/bigtable-beam-import/src/main/java/com/google/cloud/bigtable/beam/validation/HadoopHashTableSource.java:314]: There is a difference between scanEnded and workitemEnded. For scanEnded==true we close the range with stopKey; for workitemEnded==true we close the range with currentStartKey (this.key).

File: bigtable-dataflow-parent/bigtable-beam-import/src/main/java/com/google/cloud/bigtable/beam/validation/SyncTableJob.java [hunk 17]
```diff
@@ -0,0 +1,194 @@
...
+ *   mvn compile exec:java \                                                                                [20/11/18| 9:45PM]
+ *      -DmainClass=com.google.cloud.bigtable.beam.validation.SyncTableJob \
+ *      -Dexec.args="--runner=DataflowRunner \
+ *            --project=$PROJECT \
+ *            --bigtableInstanceId=$INSTANCE \
+ *            --bigtableTableId=$TABLE \
+ *            --hbaseRootDir=$HBASE_ROOT \
+ *            --snapshotName=$SNAPSHOT_NAME  \
+ *            --restoreDir=$RESTORE_DIR \
+ *            --hashTableOutputDir=$HASHTABLE_OUTPUT_DIR \
```
[the remark this answers is not in the document]
Comment (vermas2012) [reply bigtable-dataflow-parent/bigtable-beam-import/src/main/java/com/google/cloud/bigtable/beam/validation/SyncTableJob.java:56]: This PR was just created for the walkthrough on Thr and to get your high level feedback. I have not thought this through yet.

File: bigtable-dataflow-parent/bigtable-beam-import/src/main/java/com/google/cloud/bigtable/beam/validation/SyncTableJob.java [hunk 18]
```diff
@@ -0,0 +1,194 @@
...
+ * <p>Execute the following command to run the job directly:
+ *
+ * <pre>
+ *   mvn compile exec:java \                                                                                [20/11/18| 9:45PM]
+ *      -DmainClass=com.google.cloud.bigtable.beam.validation.SyncTableJob \
+ *      -Dexec.args="--runner=DataflowRunner \
+ *            --project=$PROJECT \
+ *            --bigtableInstanceId=$INSTANCE \
+ *            --bigtableTableId=$TABLE \
+ *            --hbaseRootDir=$HBASE_ROOT \
```
[the remark this answers is not in the document]
Comment (vermas2012) [reply bigtable-dataflow-parent/bigtable-beam-import/src/main/java/com/google/cloud/bigtable/beam/validation/SyncTableJob.java:53]: TBD, will clean it up after removing snapshot dependecy. 
This PR was just created for the walkthrough on Thr and to get your high level feedback. I have not thought this through yet.

File: bigtable-dataflow-parent/bigtable-beam-import/src/main/java/com/google/cloud/bigtable/beam/validation/SyncTableJob.java [hunk 19]
```diff
@@ -0,0 +1,194 @@
...
+ *            --snapshotName=$SNAPSHOT_NAME  \
```
[the remark this answers is not in the document]
Comment (vermas2012) [reply bigtable-dataflow-parent/bigtable-beam-import/src/main/java/com/google/cloud/bigtable/beam/validation/SyncTableJob.java:54]: This PR was just created for the walkthrough on Thr and to get your high level feedback. I have not thought this through yet.

**Comment by vermas2012** [comment]
>I took a rough pass reviewing. I'll probably have more comments after the holidays, but I wanted to jot some initial thoughts

This PR was just created for the walkthrough on Thursday and to feedback on very high level beam concepts. This PR is definitely not ready for this level yet. There are tests missing and I have not thought many things. The diff contains a combination of importsnapshot and sync-table.

I will close this PR(will incorporate all the feedback). I will create a new PR after import-snapshot is there, to get a cleaner diff.

File: bigtable-dataflow-parent/bigtable-beam-import/src/main/java/com/google/cloud/bigtable/beam/validation/HadoopHashTableSource.java [hunk 20]
```diff
@@ -0,0 +1,411 @@
...
+    ImmutableBytesWritable startRow;
+    ImmutableBytesWritable stopRow;
+
+    transient int numKeys = 0;
+    // Flags a state where reader.getCurrentKey() > stopRow, which means the workItem has ended.
+    transient boolean workItemEnded = false;
+    // Flags a state where reader.next() == false, which means that all the hashes have been read.
+    transient boolean scanEnded = false;
+    transient ImmutableBytesWritable cachedBatchStartKey;
+    transient ImmutableBytesWritable cachedBatchHash;
```
[the remark this answers is not in the document]
Comment (vermas2012) [reply bigtable-dataflow-parent/bigtable-beam-import/src/main/java/com/google/cloud/bigtable/beam/validation/HadoopHashTableSource.java:282]: Re: lastStartKey vs CachedStartKey, I was re-using the terminology from the HBase jobs, but yeah, 'last' is more meaningful. 

Re: Transient: I didn't see any harm in declaring them transient.

File: bigtable-dataflow-parent/bigtable-beam-import/src/main/java/com/google/cloud/bigtable/beam/validation/HadoopHashTableSource.java [hunk 21]
```diff
@@ -0,0 +1,411 @@
...
+    }
+  }
+
+  static class HashBasedReader extends BoundedReader<RangeHash> {
+
+    KeyBasedHashTableSource source;
+    TableHash.Reader reader;
+
+    ImmutableBytesWritable startRow;
+    ImmutableBytesWritable stopRow;
```
[the remark this answers is not in the document]
Comment (vermas2012) [reply bigtable-dataflow-parent/bigtable-beam-import/src/main/java/com/google/cloud/bigtable/beam/validation/HadoopHashTableSource.java:274]: done.

File: bigtable-dataflow-parent/bigtable-beam-import/src/main/java/com/google/cloud/bigtable/beam/validation/SyncTableJob.java [hunk 22]
```diff
@@ -0,0 +1,194 @@
...
+ *            --restoreDir=$RESTORE_DIR \
```
[the remark this answers is not in the document]
Comment (vermas2012) [reply bigtable-dataflow-parent/bigtable-beam-import/src/main/java/com/google/cloud/bigtable/beam/validation/SyncTableJob.java:55]: This is a consequence of direct dependency on import, this will go away when I remove all the import snapshot code.

File: bigtable-dataflow-parent/bigtable-beam-import/src/main/java/com/google/cloud/bigtable/beam/validation/SyncTableJob.java [hunk 23]
```diff
@@ -0,0 +1,194 @@
...
+        .apply("Write to file", TextIO.write().to(opts.getOutputPrefix()).withSuffix(".txt"));
+    return pipeline;
+  }
+
+  static class RangeHashToString extends SimpleFunction<RangeHash, String> {
+    @Override
+    public String apply(RangeHash input) {
+      return String.format(
+          "[%s, %s)",
+          Bytes.toStringBinary(input.startInclusive), Bytes.toStringBinary(input.endExclusive));
```
[the remark this answers is not in the document]
Comment (vermas2012) [reply bigtable-dataflow-parent/bigtable-beam-import/src/main/java/com/google/cloud/bigtable/beam/validation/SyncTableJob.java:191]: Ack, I haven't given a lot of thought on this. This was just for my testing.

File: bigtable-dataflow-parent/bigtable-beam-import/src/main/java/com/google/cloud/bigtable/beam/validation/TableHashWrapper.java [hunk 24]
```diff
@@ -0,0 +1,82 @@
...
+    this.hash = hash;
+  }
+
+  public static TableHashWrapper create(SerializableConfiguration conf, String hashTableOutputDir) {
+    TableHashWrapper tableHashWrapper;
+    try {
+      tableHashWrapper =
+          new TableHashWrapper(TableHash.read(conf.get(), new Path(hashTableOutputDir)));
+    } catch (IOException e) {
+      throw new RuntimeException("Failed to read HashTable's output via TableHash.read", e);
```
[the remark this answers is not in the document]
Comment (vermas2012) [reply bigtable-dataflow-parent/bigtable-beam-import/src/main/java/com/google/cloud/bigtable/beam/validation/TableHashWrapper.java:45]: old habit of throwing only Runtime exceptions.

File: bigtable-dataflow-parent/bigtable-beam-import/src/main/resources/log4j.properties (hunk 1 above)
Comment (vermas2012) [reply bigtable-dataflow-parent/bigtable-beam-import/src/main/resources/log4j.properties:26]: Added by import PR. Not part of the synctable PR.

File: pom.xml (hunk 3 above)
Comment (vermas2012) [reply pom.xml:82]: Will be done in import snapshot PR.

File: bigtable-dataflow-parent/bigtable-beam-import/src/main/java/org/apache/hadoop/hbase/mapreduce/BigtableTableHashAccessor.java [hunk 25]
```diff
@@ -0,0 +1,53 @@
...
+  public static ImmutableBytesWritable getStopRow(TableHash hash) {
+    return new ImmutableBytesWritable(hash.stopRow);
+  }
+
+  public static Scan getScan(TableHash hash) throws IOException {
+    return hash.initScan();
+  }
+
+  // Wrapper to access ResultHasher from ComputeAndValidateHashFromBigtableDoFn.
+  public static class BigtableResultHasher extends ResultHasher {}
```
[the remark this answers is not in the document]
Comment (vermas2012) [reply bigtable-dataflow-parent/bigtable-beam-import/src/main/java/org/apache/hadoop/hbase/mapreduce/BigtableTableHashAccessor.java:52]: done.

File: bigtable-dataflow-parent/bigtable-beam-import/src/test/java/com/google/cloud/bigtable/beam/validation/HadoopHashBasedReaderTest.java (hunk 2 above)
Comment (vermas2012) [reply bigtable-dataflow-parent/bigtable-beam-import/src/test/java/com/google/cloud/bigtable/beam/validation/HadoopHashBasedReaderTest.java:41]: Here, all the required methods are public, its just the class that is package private. So we are highjacking the whole class. But I have made the changes to keep the 2 classes consistent.

File: bigtable-dataflow-parent/bigtable-beam-import/src/main/java/org/apache/hadoop/hbase/mapreduce/BigtableTableHashAccessor.java (hunk 25 above)
[the remark this answers is not in the document]
Comment (vermas2012) [reply bigtable-dataflow-parent/bigtable-beam-import/src/main/java/org/apache/hadoop/hbase/mapreduce/BigtableTableHashAccessor.java:52]: Here, all the required methods are public, its just the class that is package private. So we are highjacking the whole class. But I have made the changes to keep the 2 classes consistent.

File: bigtable-dataflow-parent/bigtable-beam-import/src/main/java/com/google/cloud/bigtable/beam/validation/HadoopHashTableSource.java [hunk 26]
```diff
@@ -0,0 +1,411 @@
...
+
+/**
+ * A beam source to read output of Hadoop HashTable job. The source creates 1 workitem per HashTable
+ * data file and emits a row-range/hash pair.
+ */
+public class HadoopHashTableSource extends BoundedSource<RangeHash> {
+
+  static interface SerializableSupplier<T> extends Serializable, Supplier<T> {}
+
+  @DefaultCoder(AvroCoder.class)
```
[the remark this answers is not in the document]
Comment (vermas2012) [reply bigtable-dataflow-parent/bigtable-beam-import/src/main/java/com/google/cloud/bigtable/beam/validation/HadoopHashTableSource.java:52]: Source needs to store the start and end. We need to store that in a serializable form, so I have used byte[] for that. 

For everything else, I am using a ImmutableBytesWritable with AvroCoder. I am trying to use provided coders as much as possible. Also, the hashtable output that we are reading is supposed to be much smaller than the table, so we are not reading TiBs of data here.

File: bigtable-dataflow-parent/bigtable-beam-import/src/main/java/com/google/cloud/bigtable/beam/validation/HadoopHashTableSource.java [hunk 27]
```diff
@@ -0,0 +1,411 @@
...
+  @Override
+  public long getEstimatedSizeBytes(PipelineOptions options) throws Exception {
+    // HashTable data files don't expose a method to estimate size or lineCount.
+    return 0;
+  }
+
+  @Override
+  public BoundedReader<RangeHash> createReader(PipelineOptions options) {
+    // Reader should always come from KeyBasedSource.
+    throw new IllegalArgumentException("Reader can't be created from HadoopHashTableSource.");
```
[the remark this answers is not in the document]
Comment (vermas2012) [reply bigtable-dataflow-parent/bigtable-beam-import/src/main/java/com/google/cloud/bigtable/beam/validation/HadoopHashTableSource.java:207]: I used a similar format as CloudBigtableIO source, the dynamic parameters made it complicated. Can't do hash.getStartRow in constructor as hash is coming from a ValueProvider<> not available at the runtime. 

I merged the 2 classes, the constructors are not as clean as I would have liked, due to the valueproviders, but I think it will work.

File: bigtable-dataflow-parent/bigtable-beam-import/src/main/java/com/google/cloud/bigtable/beam/validation/HadoopHashTableSource.java [hunk 28]
```diff
@@ -0,0 +1,411 @@
...
+    private final String stopRow;
+
+    public KeyBasedHashTableSource(
+        SerializableConfiguration conf,
+        String hadoopHashTableOutputDir,
+        ImmutableBytesWritable startRow,
+        ImmutableBytesWritable stopRow) {
+      super(conf, StaticValueProvider.of(hadoopHashTableOutputDir));
+      this.startRow = encodeImmutableBytesWritable(startRow);
+      this.stopRow = encodeImmutableBytesWritable(stopRow);
```
[the remark this answers is not in the document]
Comment (vermas2012) [reply bigtable-dataflow-parent/bigtable-beam-import/src/main/java/com/google/cloud/bigtable/beam/validation/HadoopHashTableSource.java:223]: See my earlier comment for the reasoning. I have also consolidated them.

File: bigtable-dataflow-parent/bigtable-beam-import/src/main/java/com/google/cloud/bigtable/beam/validation/SyncTableJob.java (hunk 19 above)
[the remark this answers is not in the document]
Comment (vermas2012) [reply bigtable-dataflow-parent/bigtable-beam-import/src/main/java/com/google/cloud/bigtable/beam/validation/SyncTableJob.java:54]: removed all of it.

File: bigtable-dataflow-parent/bigtable-beam-import/src/main/java/com/google/cloud/bigtable/beam/validation/SyncTableJob.java [hunk 29]
```diff
@@ -0,0 +1,194 @@
...
+            opts.getSnapshotName(),
+            opts.getRestoreDir());
+    Pipeline pipeline = Pipeline.create(Utils.tweakOptions(opts));
+
+    pipeline
+        .apply(
+            "Read HBase HashTable output",
+            Read.from(
+                new HadoopHashTableSource(
+                    new SerializableConfiguration(conf.getHbaseConf()),
```
[the remark this answers is not in the document]
Comment (vermas2012) [reply bigtable-dataflow-parent/bigtable-beam-import/src/main/java/com/google/cloud/bigtable/beam/validation/SyncTableJob.java:173]: done.
